/* { dg-do preprocess } */
/* { dg-options "-std=c11 -pedantic" { target c } } */
/* { dg-options "-std=c++11 -pedantic" { target c++ } } */

Â¨

BÌ€

Ì€ /* { dg-error "not valid at the start of an identifier" } */

AÌ€ /* { dg-warning "not in NFC" } */

ğ€€
ğŸ¿½
ó¡ˆ´
