diff -uNr orp.orig/base_natives/gnu_classpath/gnu_specific.cpp orp/base_natives/gnu_classpath/gnu_specific.cpp
--- orp.orig/base_natives/gnu_classpath/gnu_specific.cpp	Fri Nov 16 04:14:38 2001
+++ orp/base_natives/gnu_classpath/gnu_specific.cpp	Tue Nov 27 23:19:42 2001
@@ -298,7 +298,6 @@
 //libpath.length != 0; but it's ok in reality without libpath, for
 //system can lookup needed libs in default pathes.
 
-	"java/lang/Runtime.loadLibrary(Ljava/lang/String;)V",
 #endif
 
 //GNU Classpath has intern, we keep it here for efficiency
@@ -310,6 +309,7 @@
 
 #endif //NON_ORP_NATIVE_LIBS
 
+	"java/lang/Runtime.loadLibrary(Ljava/lang/String;)V",
 }; //methods_with_native_impl
 
 static int num_methods_with_native_impl = sizeof(methods_with_native_impl) / sizeof(methods_with_native_impl[0]);
