include!(;

struct Baz {

impl Bar for


fn main() { )// { dg-error "unexpected closing delimiter .\\)." }
             // { dg-error "unexpected token .end of file. - expecting closing delimiter .\}. .for a delimited token tree." "" { target *-*-* } .+2 }
             // { dg-error "unexpected token .end of file. - expecting closing delimiter .\\). .for a macro invocation semi." "" { target *-*-* } .+1 }
