// P1949R7
// { dg-do compile }
// { dg-options "" }

bool ğŸ‘· = true;
bool ğŸ‘·â€â™€ = false;	// { dg-error "is not valid in an identifier" }
int â° = 0;	// { dg-error "is not valid in an identifier" }
int ğŸ• = 0;
int â˜  = 0;	// { dg-error "is not valid in an identifier" }
int ğŸ’€ = 0;
int âœ‹ = 0;	// { dg-error "is not valid in an identifier" }
int ğŸ‘Š = 0;
int âœˆ = 0;	// { dg-error "is not valid in an identifier" }
int ğŸš€ = 0;
int â˜¹ = 0;	// { dg-error "is not valid in an identifier" }
int ğŸ˜€ = 0;
struct E {};
class ğŸ’© : public E {};
