// P1949R7
// { dg-do compile }
// { dg-options "-pedantic-errors" }

bool ğŸ‘· = true;	// { dg-error "is not valid in an identifier" "" { target { c++98_only || c++23 } } }
bool ğŸ‘·â€â™€ = false;	// { dg-error "is not valid in an identifier" }
int â° = 0;	// { dg-error "is not valid in an identifier" }
int ğŸ• = 0;	// { dg-error "is not valid in an identifier" "" { target { c++98_only || c++23 } } }
int â˜  = 0;	// { dg-error "is not valid in an identifier" }
int ğŸ’€ = 0;	// { dg-error "is not valid in an identifier" "" { target { c++98_only || c++23 } } }
int âœ‹ = 0;	// { dg-error "is not valid in an identifier" }
int ğŸ‘Š = 0;	// { dg-error "is not valid in an identifier" "" { target { c++98_only || c++23 } } }
int âœˆ = 0;	// { dg-error "is not valid in an identifier" }
int ğŸš€ = 0;	// { dg-error "is not valid in an identifier" "" { target { c++98_only || c++23 } } }
int â˜¹ = 0;	// { dg-error "is not valid in an identifier" }
int ğŸ˜€ = 0;	// { dg-error "is not valid in an identifier" "" { target { c++98_only || c++23 } } }
struct E {};
class ğŸ’© : public E {};	// { dg-error "is not valid in an identifier" "" { target { c++98_only || c++23 } } }
