/* { dg-do compile } */
/* { dg-options "-O3 -march=rv32gc_zve32x_zvl64b -mabi=ilp32d" } */

void f___rvv_int8mf8x2_t () {__rvv_int8mf8x2_t t;} /* { dg-error {unknown type name '__rvv_int8mf8x2_t'} } */
void f___rvv_uint8mf8x2_t () {__rvv_uint8mf8x2_t t;} /* { dg-error {unknown type name '__rvv_uint8mf8x2_t'} } */
void f___rvv_int8mf8x3_t () {__rvv_int8mf8x3_t t;} /* { dg-error {unknown type name '__rvv_int8mf8x3_t'} } */
void f___rvv_uint8mf8x3_t () {__rvv_uint8mf8x3_t t;} /* { dg-error {unknown type name '__rvv_uint8mf8x3_t'} } */
void f___rvv_int8mf8x4_t () {__rvv_int8mf8x4_t t;} /* { dg-error {unknown type name '__rvv_int8mf8x4_t'} } */
void f___rvv_uint8mf8x4_t () {__rvv_uint8mf8x4_t t;} /* { dg-error {unknown type name '__rvv_uint8mf8x4_t'} } */
void f___rvv_int8mf8x5_t () {__rvv_int8mf8x5_t t;} /* { dg-error {unknown type name '__rvv_int8mf8x5_t'} } */
void f___rvv_uint8mf8x5_t () {__rvv_uint8mf8x5_t t;} /* { dg-error {unknown type name '__rvv_uint8mf8x5_t'} } */
void f___rvv_int8mf8x6_t () {__rvv_int8mf8x6_t t;} /* { dg-error {unknown type name '__rvv_int8mf8x6_t'} } */
void f___rvv_uint8mf8x6_t () {__rvv_uint8mf8x6_t t;} /* { dg-error {unknown type name '__rvv_uint8mf8x6_t'} } */
void f___rvv_int8mf8x7_t () {__rvv_int8mf8x7_t t;} /* { dg-error {unknown type name '__rvv_int8mf8x7_t'} } */
void f___rvv_uint8mf8x7_t () {__rvv_uint8mf8x7_t t;} /* { dg-error {unknown type name '__rvv_uint8mf8x7_t'} } */
void f___rvv_int8mf8x8_t () {__rvv_int8mf8x8_t t;} /* { dg-error {unknown type name '__rvv_int8mf8x8_t'} } */
void f___rvv_uint8mf8x8_t () {__rvv_uint8mf8x8_t t;} /* { dg-error {unknown type name '__rvv_uint8mf8x8_t'} } */
void f___rvv_int8mf4x2_t () {__rvv_int8mf4x2_t t;}
void f___rvv_uint8mf4x2_t () {__rvv_uint8mf4x2_t t;}
void f___rvv_int8mf4x3_t () {__rvv_int8mf4x3_t t;}
void f___rvv_uint8mf4x3_t () {__rvv_uint8mf4x3_t t;}
void f___rvv_int8mf4x4_t () {__rvv_int8mf4x4_t t;}
void f___rvv_uint8mf4x4_t () {__rvv_uint8mf4x4_t t;}
void f___rvv_int8mf4x5_t () {__rvv_int8mf4x5_t t;}
void f___rvv_uint8mf4x5_t () {__rvv_uint8mf4x5_t t;}
void f___rvv_int8mf4x6_t () {__rvv_int8mf4x6_t t;}
void f___rvv_uint8mf4x6_t () {__rvv_uint8mf4x6_t t;}
void f___rvv_int8mf4x7_t () {__rvv_int8mf4x7_t t;}
void f___rvv_uint8mf4x7_t () {__rvv_uint8mf4x7_t t;}
void f___rvv_int8mf4x8_t () {__rvv_int8mf4x8_t t;}
void f___rvv_uint8mf4x8_t () {__rvv_uint8mf4x8_t t;}
void f___rvv_int8mf2x2_t () {__rvv_int8mf2x2_t t;}
void f___rvv_uint8mf2x2_t () {__rvv_uint8mf2x2_t t;}
void f___rvv_int8mf2x3_t () {__rvv_int8mf2x3_t t;}
void f___rvv_uint8mf2x3_t () {__rvv_uint8mf2x3_t t;}
void f___rvv_int8mf2x4_t () {__rvv_int8mf2x4_t t;}
void f___rvv_uint8mf2x4_t () {__rvv_uint8mf2x4_t t;}
void f___rvv_int8mf2x5_t () {__rvv_int8mf2x5_t t;}
void f___rvv_uint8mf2x5_t () {__rvv_uint8mf2x5_t t;}
void f___rvv_int8mf2x6_t () {__rvv_int8mf2x6_t t;}
void f___rvv_uint8mf2x6_t () {__rvv_uint8mf2x6_t t;}
void f___rvv_int8mf2x7_t () {__rvv_int8mf2x7_t t;}
void f___rvv_uint8mf2x7_t () {__rvv_uint8mf2x7_t t;}
void f___rvv_int8mf2x8_t () {__rvv_int8mf2x8_t t;}
void f___rvv_uint8mf2x8_t () {__rvv_uint8mf2x8_t t;}
void f___rvv_int8m1x2_t () {__rvv_int8m1x2_t t;}
void f___rvv_uint8m1x2_t () {__rvv_uint8m1x2_t t;}
void f___rvv_int8m1x3_t () {__rvv_int8m1x3_t t;}
void f___rvv_uint8m1x3_t () {__rvv_uint8m1x3_t t;}
void f___rvv_int8m1x4_t () {__rvv_int8m1x4_t t;}
void f___rvv_uint8m1x4_t () {__rvv_uint8m1x4_t t;}
void f___rvv_int8m1x5_t () {__rvv_int8m1x5_t t;}
void f___rvv_uint8m1x5_t () {__rvv_uint8m1x5_t t;}
void f___rvv_int8m1x6_t () {__rvv_int8m1x6_t t;}
void f___rvv_uint8m1x6_t () {__rvv_uint8m1x6_t t;}
void f___rvv_int8m1x7_t () {__rvv_int8m1x7_t t;}
void f___rvv_uint8m1x7_t () {__rvv_uint8m1x7_t t;}
void f___rvv_int8m1x8_t () {__rvv_int8m1x8_t t;}
void f___rvv_uint8m1x8_t () {__rvv_uint8m1x8_t t;}
void f___rvv_int8m2x2_t () {__rvv_int8m2x2_t t;}
void f___rvv_uint8m2x2_t () {__rvv_uint8m2x2_t t;}
void f___rvv_int8m2x3_t () {__rvv_int8m2x3_t t;}
void f___rvv_uint8m2x3_t () {__rvv_uint8m2x3_t t;}
void f___rvv_int8m2x4_t () {__rvv_int8m2x4_t t;}
void f___rvv_uint8m2x4_t () {__rvv_uint8m2x4_t t;}
void f___rvv_int8m4x2_t () {__rvv_int8m4x2_t t;}
void f___rvv_uint8m4x2_t () {__rvv_uint8m4x2_t t;}
void f___rvv_int16mf4x2_t () {__rvv_int16mf4x2_t t;} /* { dg-error {unknown type name '__rvv_int16mf4x2_t'} } */
void f___rvv_uint16mf4x2_t () {__rvv_uint16mf4x2_t t;} /* { dg-error {unknown type name '__rvv_uint16mf4x2_t'} } */
void f___rvv_int16mf4x3_t () {__rvv_int16mf4x3_t t;} /* { dg-error {unknown type name '__rvv_int16mf4x3_t'} } */
void f___rvv_uint16mf4x3_t () {__rvv_uint16mf4x3_t t;} /* { dg-error {unknown type name '__rvv_uint16mf4x3_t'} } */
void f___rvv_int16mf4x4_t () {__rvv_int16mf4x4_t t;} /* { dg-error {unknown type name '__rvv_int16mf4x4_t'} } */
void f___rvv_uint16mf4x4_t () {__rvv_uint16mf4x4_t t;} /* { dg-error {unknown type name '__rvv_uint16mf4x4_t'} } */
void f___rvv_int16mf4x5_t () {__rvv_int16mf4x5_t t;} /* { dg-error {unknown type name '__rvv_int16mf4x5_t'} } */
void f___rvv_uint16mf4x5_t () {__rvv_uint16mf4x5_t t;} /* { dg-error {unknown type name '__rvv_uint16mf4x5_t'} } */
void f___rvv_int16mf4x6_t () {__rvv_int16mf4x6_t t;} /* { dg-error {unknown type name '__rvv_int16mf4x6_t'} } */
void f___rvv_uint16mf4x6_t () {__rvv_uint16mf4x6_t t;} /* { dg-error {unknown type name '__rvv_uint16mf4x6_t'} } */
void f___rvv_int16mf4x7_t () {__rvv_int16mf4x7_t t;} /* { dg-error {unknown type name '__rvv_int16mf4x7_t'} } */
void f___rvv_uint16mf4x7_t () {__rvv_uint16mf4x7_t t;} /* { dg-error {unknown type name '__rvv_uint16mf4x7_t'} } */
void f___rvv_int16mf4x8_t () {__rvv_int16mf4x8_t t;} /* { dg-error {unknown type name '__rvv_int16mf4x8_t'} } */
void f___rvv_uint16mf4x8_t () {__rvv_uint16mf4x8_t t;} /* { dg-error {unknown type name '__rvv_uint16mf4x8_t'} } */
void f___rvv_int16mf2x2_t () {__rvv_int16mf2x2_t t;}
void f___rvv_uint16mf2x2_t () {__rvv_uint16mf2x2_t t;}
void f___rvv_int16mf2x3_t () {__rvv_int16mf2x3_t t;}
void f___rvv_uint16mf2x3_t () {__rvv_uint16mf2x3_t t;}
void f___rvv_int16mf2x4_t () {__rvv_int16mf2x4_t t;}
void f___rvv_uint16mf2x4_t () {__rvv_uint16mf2x4_t t;}
void f___rvv_int16mf2x5_t () {__rvv_int16mf2x5_t t;}
void f___rvv_uint16mf2x5_t () {__rvv_uint16mf2x5_t t;}
void f___rvv_int16mf2x6_t () {__rvv_int16mf2x6_t t;}
void f___rvv_uint16mf2x6_t () {__rvv_uint16mf2x6_t t;}
void f___rvv_int16mf2x7_t () {__rvv_int16mf2x7_t t;}
void f___rvv_uint16mf2x7_t () {__rvv_uint16mf2x7_t t;}
void f___rvv_int16mf2x8_t () {__rvv_int16mf2x8_t t;}
void f___rvv_uint16mf2x8_t () {__rvv_uint16mf2x8_t t;}
void f___rvv_int16m1x2_t () {__rvv_int16m1x2_t t;}
void f___rvv_uint16m1x2_t () {__rvv_uint16m1x2_t t;}
void f___rvv_int16m1x3_t () {__rvv_int16m1x3_t t;}
void f___rvv_uint16m1x3_t () {__rvv_uint16m1x3_t t;}
void f___rvv_int16m1x4_t () {__rvv_int16m1x4_t t;}
void f___rvv_uint16m1x4_t () {__rvv_uint16m1x4_t t;}
void f___rvv_int16m1x5_t () {__rvv_int16m1x5_t t;}
void f___rvv_uint16m1x5_t () {__rvv_uint16m1x5_t t;}
void f___rvv_int16m1x6_t () {__rvv_int16m1x6_t t;}
void f___rvv_uint16m1x6_t () {__rvv_uint16m1x6_t t;}
void f___rvv_int16m1x7_t () {__rvv_int16m1x7_t t;}
void f___rvv_uint16m1x7_t () {__rvv_uint16m1x7_t t;}
void f___rvv_int16m1x8_t () {__rvv_int16m1x8_t t;}
void f___rvv_uint16m1x8_t () {__rvv_uint16m1x8_t t;}
void f___rvv_int16m2x2_t () {__rvv_int16m2x2_t t;}
void f___rvv_uint16m2x2_t () {__rvv_uint16m2x2_t t;}
void f___rvv_int16m2x3_t () {__rvv_int16m2x3_t t;}
void f___rvv_uint16m2x3_t () {__rvv_uint16m2x3_t t;}
void f___rvv_int16m2x4_t () {__rvv_int16m2x4_t t;}
void f___rvv_uint16m2x4_t () {__rvv_uint16m2x4_t t;}
void f___rvv_int16m4x2_t () {__rvv_int16m4x2_t t;}
void f___rvv_uint16m4x2_t () {__rvv_uint16m4x2_t t;}
void f___rvv_int32mf2x2_t () {__rvv_int32mf2x2_t t;} /* { dg-error {unknown type name '__rvv_int32mf2x2_t'} } */
void f___rvv_uint32mf2x2_t () {__rvv_uint32mf2x2_t t;} /* { dg-error {unknown type name '__rvv_uint32mf2x2_t'} } */
void f___rvv_int32mf2x3_t () {__rvv_int32mf2x3_t t;} /* { dg-error {unknown type name '__rvv_int32mf2x3_t'} } */
void f___rvv_uint32mf2x3_t () {__rvv_uint32mf2x3_t t;} /* { dg-error {unknown type name '__rvv_uint32mf2x3_t'} } */
void f___rvv_int32mf2x4_t () {__rvv_int32mf2x4_t t;} /* { dg-error {unknown type name '__rvv_int32mf2x4_t'} } */
void f___rvv_uint32mf2x4_t () {__rvv_uint32mf2x4_t t;} /* { dg-error {unknown type name '__rvv_uint32mf2x4_t'} } */
void f___rvv_int32mf2x5_t () {__rvv_int32mf2x5_t t;} /* { dg-error {unknown type name '__rvv_int32mf2x5_t'} } */
void f___rvv_uint32mf2x5_t () {__rvv_uint32mf2x5_t t;} /* { dg-error {unknown type name '__rvv_uint32mf2x5_t'} } */
void f___rvv_int32mf2x6_t () {__rvv_int32mf2x6_t t;} /* { dg-error {unknown type name '__rvv_int32mf2x6_t'} } */
void f___rvv_uint32mf2x6_t () {__rvv_uint32mf2x6_t t;} /* { dg-error {unknown type name '__rvv_uint32mf2x6_t'} } */
void f___rvv_int32mf2x7_t () {__rvv_int32mf2x7_t t;} /* { dg-error {unknown type name '__rvv_int32mf2x7_t'} } */
void f___rvv_uint32mf2x7_t () {__rvv_uint32mf2x7_t t;} /* { dg-error {unknown type name '__rvv_uint32mf2x7_t'} } */
void f___rvv_int32mf2x8_t () {__rvv_int32mf2x8_t t;} /* { dg-error {unknown type name '__rvv_int32mf2x8_t'} } */
void f___rvv_uint32mf2x8_t () {__rvv_uint32mf2x8_t t;} /* { dg-error {unknown type name '__rvv_uint32mf2x8_t'} } */
void f___rvv_int32m1x2_t () {__rvv_int32m1x2_t t;}
void f___rvv_uint32m1x2_t () {__rvv_uint32m1x2_t t;}
void f___rvv_int32m1x3_t () {__rvv_int32m1x3_t t;}
void f___rvv_uint32m1x3_t () {__rvv_uint32m1x3_t t;}
void f___rvv_int32m1x4_t () {__rvv_int32m1x4_t t;}
void f___rvv_uint32m1x4_t () {__rvv_uint32m1x4_t t;}
void f___rvv_int32m1x5_t () {__rvv_int32m1x5_t t;}
void f___rvv_uint32m1x5_t () {__rvv_uint32m1x5_t t;}
void f___rvv_int32m1x6_t () {__rvv_int32m1x6_t t;}
void f___rvv_uint32m1x6_t () {__rvv_uint32m1x6_t t;}
void f___rvv_int32m1x7_t () {__rvv_int32m1x7_t t;}
void f___rvv_uint32m1x7_t () {__rvv_uint32m1x7_t t;}
void f___rvv_int32m1x8_t () {__rvv_int32m1x8_t t;}
void f___rvv_uint32m1x8_t () {__rvv_uint32m1x8_t t;}
void f___rvv_int32m2x2_t () {__rvv_int32m2x2_t t;}
void f___rvv_uint32m2x2_t () {__rvv_uint32m2x2_t t;}
void f___rvv_int32m2x3_t () {__rvv_int32m2x3_t t;}
void f___rvv_uint32m2x3_t () {__rvv_uint32m2x3_t t;}
void f___rvv_int32m2x4_t () {__rvv_int32m2x4_t t;}
void f___rvv_uint32m2x4_t () {__rvv_uint32m2x4_t t;}
void f___rvv_int32m4x2_t () {__rvv_int32m4x2_t t;}
void f___rvv_uint32m4x2_t () {__rvv_uint32m4x2_t t;}
void f___rvv_int64m1x2_t () {__rvv_int64m1x2_t t;} /* { dg-error {unknown type name '__rvv_int64m1x2_t'} } */
void f___rvv_uint64m1x2_t () {__rvv_uint64m1x2_t t;} /* { dg-error {unknown type name '__rvv_uint64m1x2_t'} } */
void f___rvv_int64m1x3_t () {__rvv_int64m1x3_t t;} /* { dg-error {unknown type name '__rvv_int64m1x3_t'} } */
void f___rvv_uint64m1x3_t () {__rvv_uint64m1x3_t t;} /* { dg-error {unknown type name '__rvv_uint64m1x3_t'} } */
void f___rvv_int64m1x4_t () {__rvv_int64m1x4_t t;} /* { dg-error {unknown type name '__rvv_int64m1x4_t'} } */
void f___rvv_uint64m1x4_t () {__rvv_uint64m1x4_t t;} /* { dg-error {unknown type name '__rvv_uint64m1x4_t'} } */
void f___rvv_int64m1x5_t () {__rvv_int64m1x5_t t;} /* { dg-error {unknown type name '__rvv_int64m1x5_t'} } */
void f___rvv_uint64m1x5_t () {__rvv_uint64m1x5_t t;} /* { dg-error {unknown type name '__rvv_uint64m1x5_t'} } */
void f___rvv_int64m1x6_t () {__rvv_int64m1x6_t t;} /* { dg-error {unknown type name '__rvv_int64m1x6_t'} } */
void f___rvv_uint64m1x6_t () {__rvv_uint64m1x6_t t;} /* { dg-error {unknown type name '__rvv_uint64m1x6_t'} } */
void f___rvv_int64m1x7_t () {__rvv_int64m1x7_t t;} /* { dg-error {unknown type name '__rvv_int64m1x7_t'} } */
void f___rvv_uint64m1x7_t () {__rvv_uint64m1x7_t t;} /* { dg-error {unknown type name '__rvv_uint64m1x7_t'} } */
void f___rvv_int64m1x8_t () {__rvv_int64m1x8_t t;} /* { dg-error {unknown type name '__rvv_int64m1x8_t'} } */
void f___rvv_uint64m1x8_t () {__rvv_uint64m1x8_t t;} /* { dg-error {unknown type name '__rvv_uint64m1x8_t'} } */
void f___rvv_int64m2x2_t () {__rvv_int64m2x2_t t;} /* { dg-error {unknown type name '__rvv_int64m2x2_t'} } */
void f___rvv_uint64m2x2_t () {__rvv_uint64m2x2_t t;} /* { dg-error {unknown type name '__rvv_uint64m2x2_t'} } */
void f___rvv_int64m2x3_t () {__rvv_int64m2x3_t t;} /* { dg-error {unknown type name '__rvv_int64m2x3_t'} } */
void f___rvv_uint64m2x3_t () {__rvv_uint64m2x3_t t;} /* { dg-error {unknown type name '__rvv_uint64m2x3_t'} } */
void f___rvv_int64m2x4_t () {__rvv_int64m2x4_t t;} /* { dg-error {unknown type name '__rvv_int64m2x4_t'} } */
void f___rvv_uint64m2x4_t () {__rvv_uint64m2x4_t t;} /* { dg-error {unknown type name '__rvv_uint64m2x4_t'} } */
void f___rvv_int64m4x2_t () {__rvv_int64m4x2_t t;} /* { dg-error {unknown type name '__rvv_int64m4x2_t'} } */
void f___rvv_uint64m4x2_t () {__rvv_uint64m4x2_t t;} /* { dg-error {unknown type name '__rvv_uint64m4x2_t'} } */
void f___rvv_float32mf2x2_t () {__rvv_float32mf2x2_t t;} /* { dg-error {unknown type name '__rvv_float32mf2x2_t'} } */
void f___rvv_float32mf2x3_t () {__rvv_float32mf2x3_t t;} /* { dg-error {unknown type name '__rvv_float32mf2x3_t'} } */
void f___rvv_float32mf2x4_t () {__rvv_float32mf2x4_t t;} /* { dg-error {unknown type name '__rvv_float32mf2x4_t'} } */
void f___rvv_float32mf2x5_t () {__rvv_float32mf2x5_t t;} /* { dg-error {unknown type name '__rvv_float32mf2x5_t'} } */
void f___rvv_float32mf2x6_t () {__rvv_float32mf2x6_t t;} /* { dg-error {unknown type name '__rvv_float32mf2x6_t'} } */
void f___rvv_float32mf2x7_t () {__rvv_float32mf2x7_t t;} /* { dg-error {unknown type name '__rvv_float32mf2x7_t'} } */
void f___rvv_float32mf2x8_t () {__rvv_float32mf2x8_t t;} /* { dg-error {unknown type name '__rvv_float32mf2x8_t'} } */
void f___rvv_float32m1x2_t () {__rvv_float32m1x2_t t;} /* { dg-error {unknown type name '__rvv_float32m1x2_t'} } */
void f___rvv_float32m1x3_t () {__rvv_float32m1x3_t t;} /* { dg-error {unknown type name '__rvv_float32m1x3_t'} } */
void f___rvv_float32m1x4_t () {__rvv_float32m1x4_t t;} /* { dg-error {unknown type name '__rvv_float32m1x4_t'} } */
void f___rvv_float32m1x5_t () {__rvv_float32m1x5_t t;} /* { dg-error {unknown type name '__rvv_float32m1x5_t'} } */
void f___rvv_float32m1x6_t () {__rvv_float32m1x6_t t;} /* { dg-error {unknown type name '__rvv_float32m1x6_t'} } */
void f___rvv_float32m1x7_t () {__rvv_float32m1x7_t t;} /* { dg-error {unknown type name '__rvv_float32m1x7_t'} } */
void f___rvv_float32m1x8_t () {__rvv_float32m1x8_t t;} /* { dg-error {unknown type name '__rvv_float32m1x8_t'} } */
void f___rvv_float32m2x2_t () {__rvv_float32m2x2_t t;} /* { dg-error {unknown type name '__rvv_float32m2x2_t'} } */
void f___rvv_float32m2x3_t () {__rvv_float32m2x3_t t;} /* { dg-error {unknown type name '__rvv_float32m2x3_t'} } */
void f___rvv_float32m2x4_t () {__rvv_float32m2x4_t t;} /* { dg-error {unknown type name '__rvv_float32m2x4_t'} } */
void f___rvv_float32m4x2_t () {__rvv_float32m4x2_t t;} /* { dg-error {unknown type name '__rvv_float32m4x2_t'} } */
void f___rvv_float64m1x2_t () {__rvv_float64m1x2_t t;} /* { dg-error {unknown type name '__rvv_float64m1x2_t'} } */
void f___rvv_float64m1x3_t () {__rvv_float64m1x3_t t;} /* { dg-error {unknown type name '__rvv_float64m1x3_t'} } */
void f___rvv_float64m1x4_t () {__rvv_float64m1x4_t t;} /* { dg-error {unknown type name '__rvv_float64m1x4_t'} } */
void f___rvv_float64m1x5_t () {__rvv_float64m1x5_t t;} /* { dg-error {unknown type name '__rvv_float64m1x5_t'} } */
void f___rvv_float64m1x6_t () {__rvv_float64m1x6_t t;} /* { dg-error {unknown type name '__rvv_float64m1x6_t'} } */
void f___rvv_float64m1x7_t () {__rvv_float64m1x7_t t;} /* { dg-error {unknown type name '__rvv_float64m1x7_t'} } */
void f___rvv_float64m1x8_t () {__rvv_float64m1x8_t t;} /* { dg-error {unknown type name '__rvv_float64m1x8_t'} } */
void f___rvv_float64m2x2_t () {__rvv_float64m2x2_t t;} /* { dg-error {unknown type name '__rvv_float64m2x2_t'} } */
void f___rvv_float64m2x3_t () {__rvv_float64m2x3_t t;} /* { dg-error {unknown type name '__rvv_float64m2x3_t'} } */
void f___rvv_float64m2x4_t () {__rvv_float64m2x4_t t;} /* { dg-error {unknown type name '__rvv_float64m2x4_t'} } */
void f___rvv_float64m4x2_t () {__rvv_float64m4x2_t t;} /* { dg-error {unknown type name '__rvv_float64m4x2_t'} } */
