/* { dg-do preprocess } */
/* { dg-options "-std=gnu++98" } */

ª
« /* { dg-error "not valid in an identifier" } */
¶ /* { dg-error "not valid in an identifier" } */
º
À
Ö
΄

٩ /* OK in C++ */
A٩
0º
0٩
๙ /* OK in C++ */
A๙
