// { dg-do compile }
// { dg-options "-fcheck-pointer-bounds -fchkp-first-field-has-own-bounds" }

// { dg-warning ".-fcheck-pointer-bounds. is no longer supported" "" { target *-*-* } 0 }
// { dg-warning ".-fchkp-first-field-has-own-bounds. is no longer supported" "" { target *-*-* } 0 }
